File: src/python_template_pkg/nodes/turtle_following_node.py
#!/usr/bin/env python3

import rospy
from turtlesim.srv import Spawn
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist

from python_template_pkg.pid_following import compute_cmd_vel

class TurtleFollowingNode:
    def __init__(self):
        self.load_parameters()
        self.init_pubsub()
        self.spawn_second_turtle()

    def load_parameters(self):
        self.turtle2_init_pose_x = rospy.get_param("~turtle2_init_pose_x")    
        self.turtle2_init_pose_y = rospy.get_param("~turtle2_init_pose_y")    
        self.turtle2_init_pose_theta = rospy.get_param("~turtle2_init_pose_theta")    
        self.kp_linear = rospy.get_param("~kp_linear")    
        self.kp_angular = rospy.get_param("~kp_angular")    

    def init_pubsub(self):
        self.turtle1_pose_sub = rospy.Subscriber('turtle1/pose', Pose, self.turtle1_pose_callback)
        self.turtle2_pose_sub = rospy.Subscriber('turtle2/pose', Pose, self.turtle2_pose_callback)
        self.turtle2_twist_pub = rospy.Publisher('turtle2/cmd_vel', Twist, queue_size=1)

    def spawn_second_turtle(self):
        spawn_turtle = rospy.ServiceProxy('spawn', Spawn)
        spawn_turtle(self.turtle2_init_pose_x, self.turtle2_init_pose_y, self.turtle2_init_pose_theta, 'turtle2')
        rospy.loginfo("Second turtle spawned successfully.")

    def turtle1_pose_callback(self, msg):
        self.turtle1_x = msg.x
        self.turtle1_y = msg.y
        self.turtle1_theta = msg.theta

    def turtle2_pose_callback(self, msg):
        self.turtle2_x = msg.x
        self.turtle2_y = msg.y
        self.turtle2_theta = msg.theta
        
        twist_msg = compute_cmd_vel(
            self.turtle1_x, self.turtle1_y, self.turtle2_x, self.turtle2_y, self.turtle2_theta,
            self.kp_linear, self.kp_angular
        )
        self.turtle2_twist_pub.publish(twist_msg)


if __name__ == "__main__":
    rospy.init_node("turtle_following_node")
    TurtleFollowingNode()
    rospy.spin()    

File: src/cpp_template_pkg/src/turtle_following_node.cpp
#include "cpp_template_pkg/turtle_following_node.h"
#include "cpp_template_pkg/pid_following.h"

TurtleFollowingNode::TurtleFollowingNode() : pnh_("~") 
{
    loadParameters();
    initPubSub();
    spawnSecondTurtle();
}

void TurtleFollowingNode::loadParameters() {
    pnh_.param("turtle2_init_pose_x", turtle2_init_pose_x_, 0.0);
    pnh_.param("turtle2_init_pose_y", turtle2_init_pose_y_, 0.0);
    pnh_.param("turtle2_init_pose_theta", turtle2_init_pose_theta_, 0.0);
    pnh_.param("kp_linear", kp_linear_, 1.5);
    pnh_.param("kp_angular", kp_angular_, 6.0);
}

void TurtleFollowingNode::initPubSub() {
    turtle1_pose_sub_ = nh_.subscribe("turtle1/pose", 1, &TurtleFollowingNode::turtle1PoseCallback, this);
    turtle2_pose_sub_ = nh_.subscribe("turtle2/pose", 1, &TurtleFollowingNode::turtle2PoseCallback, this);
    turtle2_twist_pub_ = nh_.advertise<geometry_msgs::Twist>("turtle2/cmd_vel", 1);
}

void TurtleFollowingNode::spawnSecondTurtle() {
    ros::ServiceClient spawn_turtle = nh_.serviceClient<turtlesim::Spawn>("spawn");
    turtlesim::Spawn srv;
    srv.request.x = turtle2_init_pose_x_;
    srv.request.y = turtle2_init_pose_y_;
    srv.request.theta = turtle2_init_pose_theta_;
    srv.request.name = "turtle2";
    if (spawn_turtle.call(srv)) {
        ROS_INFO("Second turtle spawned successfully.");
    } else {
        ROS_ERROR("Failed to spawn second turtle.");
    }
}

TurtleFollowingNode::~TurtleFollowingNode() {}

void TurtleFollowingNode::turtle1PoseCallback(const turtlesim::Pose::ConstPtr& msg) {
    turtle1_x_ = msg->x;
    turtle1_y_ = msg->y;
    turtle1_theta_ = msg->theta;
}

void TurtleFollowingNode::turtle2PoseCallback(const turtlesim::Pose::ConstPtr& msg) {
    turtle2_x_ = msg->x;
    turtle2_y_ = msg->y;
    turtle2_theta_ = msg->theta;

    geometry_msgs::Twist twist_msg = computeCmdVel(
        turtle1_x_, turtle1_y_, turtle2_x_, turtle2_y_, turtle2_theta_,
        kp_linear_, kp_angular_
    );
    turtle2_twist_pub_.publish(twist_msg);
}

int main(int argc, char** argv) 
{
    ros::init(argc, argv, "turtle_following_node");
    TurtleFollowingNode node;
    ros::spin();
    return 0;
}
